server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,321 @@
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    public void initiateAttributeDecommissioning(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener,
+        ClusterState state
+    ) {
+        /*
+         * 1. Abdicate master
+         * 2. Register attribute -> status should be set to INIT
+         * 3. Trigger weigh away for graceful decommission -> status should be set to DECOMMISSIONING
+         * 4. Once zone is weighed away -> trigger zone decommission using executor -> status should be set to DECOMMISSIONED on successful response
+         * 5. Clear voting config
+         */
+        this.clusterState = state;
+        logger.info("initiating awareness attribute [{}] decommissioning", decommissionAttribute.toString());
+        excludeDecommissionedClusterManagerNodesFromVotingConfig(decommissionAttribute, listener);
+    }
+
+    private void excludeDecommissionedClusterManagerNodesFromVotingConfig(
+        DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        final Predicate<DiscoveryNode> shouldAbdicatePredicate = discoveryNode -> nodeHasDecommissionedAttribute(discoveryNode, decommissionAttribute);
+        List<String> clusterManagerNodesToBeDecommissioned = new ArrayList<>();
+        Iterator<DiscoveryNode> clusterManagerNodesIter = clusterState.nodes().getClusterManagerNodes().valuesIt();
+        while (clusterManagerNodesIter.hasNext()) {
+            final DiscoveryNode node = clusterManagerNodesIter.next();
+            if (shouldAbdicatePredicate.test(node)) {
+                clusterManagerNodesToBeDecommissioned.add(node.getName());
+            }
+        }
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            AddVotingConfigExclusionsAction.NAME,
+            new AddVotingConfigExclusionsRequest(clusterManagerNodesToBeDecommissioned.toArray(String[]::new)),
+            new TransportResponseHandler<AddVotingConfigExclusionsResponse>() {
+                @Override
+                public void handleResponse(AddVotingConfigExclusionsResponse response) {
+                    logger.info("successfully removed decommissioned cluster manager eligible nodes [{}] from voting config, " +
+                        "proceeding to drain the decommissioned nodes", clusterManagerNodesToBeDecommissioned.toString());
+                    registerDecommissionAttribute(decommissionAttribute, listener);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.debug(new ParameterizedMessage(
+                        "failure in removing decommissioned cluster manager eligible nodes from voting config"), exp);
+                }
```

Comment:
Decommission process won't even start since voting config exclusion failed. IMO, we won't need any metadata/status update in this case. Do you see a use case where status failed due to failure in voting config exclusion will help?

Reply:
Update - Only once INIT is registered we attempt to exclude the eligible cluster manager nodes. The entire request of decommissioning is ensured that always run on master that will never be decommissioned. Also, the request is only processed when we know that quorum is maintained even after removal of nodes. The user can retry the request or clear the metadata with DELETE in any case.